### PR TITLE
Issue #820: Clean-up FLI to remove code that was causing stack traces

### LINKED ISSUE #820: FLI causes stack trace in mixed language simulations
In mixed language simulations, I have seen issues with the dynamic casting causing stack traces.  It doesn't always happen and can be totally dependent on the type of top one is using.  The problem is, the code made some assumptions about the 'parent' handle when doing the cast and that assumption isn't always holding true, i.e. the parent isn't an FLI handle, causing the cast to fail and hence a stack trace.

I believe we can get rid of the need to do the casting, which will remove the risk of the stack traces.

### PR DESCRIPTION
**Pull request by leftink**
Fixes #820 

This is a pretty large change to the FLI interface.  In mixed language simulations, I had seen stack traces caused by the casting to FliObj.  I reworked the interface to remove any requirement to cast to any FLI object.

### CODE REVIEW

File: cocotb/share/lib/fli/FliObjHdl.cpp [hunk 1]
```diff
@@ -98,21 +58,79 @@ int FliObjHdl::initialise(std::string &name, std::string &fq_name)
...
-int FliSignalObjHdl::initialise(std::string &name, std::string &fq_name)
+int FliSignalObjHdl::initialise(std::string &name, std::string &fq_name, fli_type_t fli_type)
 {
+    if (fli_type == FLI_TYPE_SIGNAL) {
+        if ((m_rising_cb = new FliSignalCbHdl(m_impl, this, GPI_RISING)) == NULL) {
+            LOG_CRITICAL("Failed to allocate memory for Rising-Edge Callback. (%s (%s))", name.c_str(), get_type_str());
+            return -1;
+        }
+        if ((m_falling_cb = new FliSignalCbHdl(m_impl, this, GPI_FALLING)) == NULL) {
+            LOG_CRITICAL("Failed to allocate memory for Fallingg-Edge Callback. (%s (%s))", name.c_str(), get_type_str());
```
Comment (cmarqu) [inline cocotb/share/lib/fli/FliObjHdl.cpp]: Small typo: "Fallingg"

Comment (leftink) [reply cocotb/share/lib/fli/FliObjHdl.cpp]: Agreed.  I pushed a fix.

File: cocotb/share/lib/fli/FliObjHdl.cpp [hunk 2]
```diff
@@ -98,21 +58,79 @@ int FliObjHdl::initialise(std::string &name, std::string &fq_name)
...
+int FliSignalObjHdl::initialise(std::string &name, std::string &fq_name, fli_type_t fli_type)
 {
+    if (fli_type == FLI_TYPE_SIGNAL) {
+        if ((m_rising_cb = new FliSignalCbHdl(m_impl, this, GPI_RISING)) == NULL) {
+            LOG_CRITICAL("Failed to allocate memory for Rising-Edge Callback. (%s (%s))", name.c_str(), get_type_str());
+            return -1;
+        }
+        if ((m_falling_cb = new FliSignalCbHdl(m_impl, this, GPI_FALLING)) == NULL) {
+            LOG_CRITICAL("Failed to allocate memory for Falling-Edge Callback. (%s (%s))", name.c_str(), get_type_str());
+            return -1;
```
Comment (eric-wieser) [inline cocotb/share/lib/fli/FliObjHdl.cpp]: `new` never returns `NULL`, these lines are pointless

Comment (leftink) [reply cocotb/share/lib/fli/FliObjHdl.cpp]: Agreed.  Updates made.

File: cocotb/share/lib/fli/FliImpl.h [hunk 3]
```diff
@@ -150,102 +150,75 @@ class FliTimedCbHdl : public FliProcessCbHdl {
...
-    FliSignalCbHdl     m_rising_cb;
-    FliSignalCbHdl     m_falling_cb;
-    FliSignalCbHdl     m_either_cb;
+    FliSignalCbHdl      *m_rising_cb;
+    FliSignalCbHdl      *m_falling_cb;
+    FliSignalCbHdl      *m_either_cb;
+    mti_GetValue         m_get_value;
+    mti_SetValue         m_set_value;
+    mti_GetArrayValue    m_get_array_value;
+    mti_GetValueIndirect m_get_value_indirect;
```
Comment (eric-wieser) [inline cocotb/share/lib/fli/FliImpl.h]: This looks like a use case for virtual functions, not function pointers and if statements

File: cocotb/share/lib/fli/FliObjHdl.cpp [hunk 4]
```diff
@@ -98,21 +58,70 @@ int FliObjHdl::initialise(std::string &name, std::string &fq_name)
...
+        m_falling_cb = new FliSignalCbHdl(m_impl, this, GPI_FALLING);
+        m_either_cb  = new FliSignalCbHdl(m_impl, this, GPI_FALLING | GPI_RISING);
+
+        m_get_value = (mti_GetValue)&mti_GetSignalValue;
+        m_set_value = (mti_SetValue)&mti_SetSignalValue;
+        m_get_array_value = (mti_GetArrayValue)&mti_GetArraySignalValue;
+        m_get_value_indirect = (mti_GetValueIndirect)&mti_GetSignalValueIndirect;
+    } else {
+        m_get_value = (mti_GetValue)&mti_GetVarValue;
+        m_set_value = (mti_SetValue)&mti_SetVarValue;
```
Comment (eric-wieser) [inline cocotb/share/lib/fli/FliObjHdl.cpp]: Why is this cast any safer than the previous `dynamic_cast`?

Also, this is C++ so you should be using `reinterpret_cast<mti_SetValue>`, which acts as a big red flag that this is dangerous.

File: cocotb/share/lib/fli/FliImpl.h (hunk 3 above)
Comment (leftink) [reply cocotb/share/lib/fli/FliImpl.h]: Actually, I disagree with that assessment.  FLI is a pain purely because it treats Regions, Signal and Variables differently.  Signals and Variables are essentially the same, just different handle type and function name.  If I went the virtual function approach there would be twice as many classes that are derived from FliValueObjHdl because you would have to have a sub-class for signals and a sub-class for variables.  The code in these classes would be identical to each other, except the name of the function be called.  That seems to open up the potential of bugs getting fixed in one and not the other. 

The function pointer approach allows the classes to not care whether it is operating on a signal or a variable, which allows for a single common code base to be maintained.

File: cocotb/share/lib/fli/FliObjHdl.cpp (hunk 4 above)
Comment (leftink) [reply cocotb/share/lib/fli/FliObjHdl.cpp]: This cast isn't very dangerous at all.  We store the Signal or Variable Handle as a void *.  The type of handle is known when the initialise() function is called and that is when the function pointers are configured.  The cast is just to allow the code to pass the `void *` instead of have to cast it back to the original type.

The previous cast was dangerous because it was casting  derivatives of the GpiObjHdl and we couldn't be sure which interface created the GpiObjHdl.  If we are creating a derivative of the FliSignalObjHdl, then we know the  FLI handle is that of a Signal or a Variable.  

I did go ahead and change the code to use the reinterpret_cast<>() format.

File: cocotb/share/lib/fli/FliImpl.h (hunk 3 above)
Comment (eric-wieser) [reply cocotb/share/lib/fli/FliImpl.h]: > The code in these classes would be identical to each other, except the name of the function be called. That seems to open up the potential of bugs getting fixed in one and not the other.

You'd make them share a base class. The idea is that instead of storing the FLI type in a variable, you store it in the type of the object itself. That way you don't have to put if statements in all the methods to switch between types.

Comment (leftink) [reply cocotb/share/lib/fli/FliImpl.h]: What if statements are you referring to?   With my approach I have removed all the if statements.  And I don't store the FLI type.

**Comment by leftink** [comment]
@eric-wieser, I have added a test and design that shows the issue is fixed.  If you try it on the current master, it reproduced the stack trace in my environment.

**Comment by leftink** [comment]
@eric-wieser, the latest was my attempt to remove the function pointers.  Hopefully this resolves any concerns you had.

**Comment by leftink** [comment]
@eric-wieser, I was curious if there is anything else holding up these changes.  I have another set of fixes for #813 that are dependent on these fixes.  I have been holding off on submitting the other PR until all issues are addressed with this PR.

**Comment by eric-wieser** [comment]
@leftink you have me confused with someone who has the ability to merge. There's a lot to look at here, and I don't really have the background to review it all anyway. Hopefully a maintainer will be able to review - if not, breaking this into smaller refectors might make it more reviewable.

**Comment by leftink** [comment]
@eric-wieser, My bad. Didn't realize you weren't one of the maintainers.

**Comment by imphil** [comment]
@leftink it's hard to review this PR since it mixes a couple of unrelated commits into the FLI change itself. Could you please open a separate PR for the unrelated issues (e.g. the test fix), and then rebase the commits in this PR to be self-contained commits (as opposed to multiple commits which show how this PR evolved).

Also please follow https://github.com/potentialventures/cocotb/blob/master/CONTRIBUTING.md#patch-requirements: keep the summary line of the commit message below 50 characters and have a longer commit message explaining the need for the change and the approach taken.

**Comment by leftink** [comment]
@imphil, per your request I have split the two test fixes into separate pull requests and squashed all the other changes to a single commit.

**Comment by eric-wieser** [comment]
> As for the void * pointers in those c-style functions, either you duplicate the code or you pass the pointer in agnostic of the type. 

You might be able to go a long way by templating the functions

> it is interfacing to a C-style API

Unless C code is actually linking against these functions, there is no reason to mark them `extern "C"`

> split the un-related commits into new pull requests

This was great, thanks for doing that - but I think the real question was whether the main piece of work can be split into smaller pieces.

File: cocotb/share/lib/fli/FliImpl.cpp [hunk 5]
```diff
@@ -106,110 +88,215 @@ bool FliImpl::isValueBoolean(mtiTypeIdT type)
     return false;
 }
 
-bool FliImpl::isTypeValue(int type)
+bool fli_is_signal(void *hdl)
+{
+    return (acc_fetch_type(hdl) == accSignal || acc_fetch_fulltype(hdl) == accAliasSignal);
```
Comment (eric-wieser) [inline cocotb/share/lib/fli/FliImpl.cpp]: An aside - where is the definition of `acc_fetch_type`? I'm struggling to even find a signature for that function online, let alone documentation.

File: cocotb/share/lib/fli/FliImpl.cpp [hunk 6]
```diff
@@ -106,110 +88,215 @@ bool FliImpl::isValueBoolean(mtiTypeIdT type)
...
+            }
+            break;
+        case MTI_TYPE_RECORD:
+            rv = GPI_STRUCTURE;
+            break;
+        default:
+            rv = GPI_UNKNOWN;
+    }
+
+    return rv;
```
[the remark this answers is not in the document]
Comment (eric-wieser) [reply cocotb/share/lib/fli/FliImpl.cpp]: Doing so also removes every break, and means the reader doesn't have to check that rv is overwritten

[the remark this answers is not in the document]
Comment (leftink) [reply cocotb/share/lib/fli/FliImpl.cpp]: Good point.  I will update.

**Comment by leftink** [comment]
@eric-wieser 
* I will think about templates, struggled with the different functions that need to be called, not just type difference.
* Good point on the functions being extern "c", that was a mistake when I was moving some things around.  I will fix.

File: cocotb/share/lib/fli/FliImpl.cpp (hunk 5 above)
Comment (leftink) [reply cocotb/share/lib/fli/FliImpl.cpp]: I believe acc_fetch_type() is part of the PLI interface, but not sure on that, no real documentation.  I believe @stuarthodgson was the first to add using that function as it is the only way to get what type handle is with more detail, i.e. constant, generic, generate loop, etc.  I doubt you will find anything on the web about the interface.

Comment (eric-wieser) [reply cocotb/share/lib/fli/FliImpl.cpp]: Is there even a copy of the header file I can obtain somewhere?

**Comment by leftink** [comment]
@imphil Hopefully these updates will suffice.  The primary thing I was unable to accomplish was trying to use templates to remove some of the code.  I tried several different approaches and either it wasn't any better or I was unable to get it to work.  If you feel that is needed, I ask that perhaps we open an Issue and someone with more template experience than I could work that.

File: cocotb/share/lib/fli/FliImpl.cpp [hunk 7]
```diff
@@ -85,13 +68,13 @@ bool FliImpl::isValueLogic(mtiTypeIdT type)
...
 
-bool FliImpl::isValueChar(mtiTypeIdT type)
+bool fli_is_char(mtiTypeIdT type)
 {
     const int NUM_ENUMS_IN_CHAR_TYPE = 256;
     return (mti_TickLength(type) == NUM_ENUMS_IN_CHAR_TYPE);
 }
 
-bool FliImpl::isValueBoolean(mtiTypeIdT type)
+bool fli_is_boolean(mtiTypeIdT type)
```
Comment (eric-wieser) [inline cocotb/share/lib/fli/FliImpl.cpp]: Is this function used anywhere but in this file? If not, these should all be within
```C++
/** local helper functions */
namespace {

}
```
Which is the C++ way of spelling static

File: cocotb/share/lib/fli/FliImpl.cpp (hunk 5 above)
Comment (leftink) [reply cocotb/share/lib/fli/FliImpl.cpp]: I am not comfortable distributing it since it is mentor's file.  So I would say unless you have Modelsim or Questa, probably not.

File: cocotb/share/lib/fli/FliImpl.cpp (hunk 7 above)
Comment (leftink) [reply cocotb/share/lib/fli/FliImpl.cpp]: Agreed.

File: cocotb/share/lib/fli/FliObjHdl.cpp [hunk 8]
```diff
@@ -217,84 +288,60 @@ int FliEnumObjHdl::set_signal_value(const long value)
...
 
     for (mtiInt32T i = 0; i < m_num_enum; i++) {
         m_enum_map[m_value_enum[i][1]] = i;  // enum is of the format 'U' or '0', etc.
     }
 
-    m_val_buff = (char*)malloc(m_num_elems+1);
-    if (!m_val_buff) {
-        LOG_CRITICAL("Unable to alloc mem for value object read buffer: ABORTING");
-    }
+    m_val_buff = new char[m_num_elems+1];
```
Comment (eric-wieser) [inline cocotb/share/lib/fli/FliObjHdl.cpp]: This is a good change, but the type of change I would expect to see in its own PR

File: cocotb/share/lib/fli/FliImpl.cpp (hunk 5 above)
Comment (eric-wieser) [reply cocotb/share/lib/fli/FliImpl.cpp]: That's fair. Can you tell me the name of the header it's defined in, so that I can try and source it from elsewhere?

File: cocotb/share/lib/fli/FliImpl.cpp [hunk 9]
```diff
@@ -38,26 +38,12 @@ extern "C" {
...
-{
-    if (GPI_DELETE != sim_finish_cb->get_call_state()) {
-        sim_finish_cb->set_call_state(GPI_DELETE);
-        if (mti_NowUpper() == 0 && mti_Now() == 0 && mti_Delta() == 0) {
-            mti_Quit();
-        } else {
-            mti_Break();
-        }
-    }
-}
```
Comment (eric-wieser) [inline cocotb/share/lib/fli/FliImpl.cpp]: In the interest of keeping this diff small, I'd err on the side of not moving this function

File: cocotb/share/lib/fli/FliImpl.cpp [hunk 10]
```diff
@@ -85,13 +71,13 @@ bool FliImpl::isValueLogic(mtiTypeIdT type)
     return false;
 }
 
-bool FliImpl::isValueChar(mtiTypeIdT type)
+bool fli_is_char(mtiTypeIdT type)
```
Comment (eric-wieser) [inline cocotb/share/lib/fli/FliImpl.cpp]: Is there any reason these actually need to be changed from the previous member functions to free functions?

I'm not opposed to changing them, I'm just trying to follow if this is unrelated tidy-up. If so, this too would make sense as a small standalone PR.

Remember it's not just about making the review easy - it's about giving people looking at the blame some idea of why things were changed.

File: cocotb/share/lib/fli/FliImpl.cpp (hunk 5 above)
Comment (leftink) [reply cocotb/share/lib/fli/FliImpl.cpp]: acc_user.h and acc_vhdl.h

File: cocotb/share/lib/fli/FliObjHdl.cpp (hunk 8 above)
Comment (leftink) [reply cocotb/share/lib/fli/FliObjHdl.cpp]: I agree, or at least on its own commit.  This change was on a separate commit prior to request to squash.

File: cocotb/share/lib/fli/FliImpl.cpp (hunk 9 above)
Comment (leftink) [reply cocotb/share/lib/fli/FliImpl.cpp]: Unless the functions are moving back to be part of the FliImpl, I disagree.  I prefer to keep class implementation code together.

File: cocotb/share/lib/fli/FliImpl.cpp (hunk 10 above)
Comment (leftink) [reply cocotb/share/lib/fli/FliImpl.cpp]: This is just some clean-up.  It was on separate commit prior to the request to squash, but given the state now, I am not looking to change unless there is compelling reason to do so.

**Comment by ktbarrett** [comment]
Superseded by #2776.
